### ISSUE TITLE
Replace __align fields with #[repr(align(x))]

### ISSUE THREAD

**Issue by faern** [rust-lang/libc#1042]
This has been briefly touched in #963 and #58 but this issue intend to discuss it from a different angle.

When I grep in the libc repository I find 123 instances of `__align:`. The main problem is that instantiating a struct that uses this way to achieve alignment requires:
```rust
let mut foo: libc::some_struct = unsafe { mem::uninitialized() };
foo.field = my_value;
```
Which has many downsides, mainly
* needing `unsafe`
* has to be assigned to a mutable variable
* and can't be done in a constant expression.

`#repr(align(x))]` was introduced in Rust [1.25](https://blog.rust-lang.org/2018/03/29/Rust-1.25.html), which at the time of writing this is only two releases behind current stable. Thus I understand we probably can't bump the required rustc version for libc that high yet. But I could not find any information about how backwards compatible libc intends to be, going forward. Since #972 bumped the required rustc version from 1.0 to 1.13 without much fanfare or breaking version bumping etc I guess there is some unofficial point where even libc don't care about your old compiler anymore.

Switching to `#repr(align(x))]` would, among removing the need for a lot of unsafe, help in making more things constant expressions. For example this issue where we need to create an `in6_addr` in an associated constant: https://github.com/rust-lang/rust/issues/44582

Would it count as a breaking change to remove the `__align` fields? It would change the struct, but not the public part of it, so I guess not. Any code using the `mem::uninitialized()` dance should still work exactly the same I think.

I'm playing with the thought to somehow make this work in Edition 2018. But I guess compiling a crate with `edition = "2018"` that has edition 2015 dependencies only compiles libc once, even if crates from both editions depend on it? If it is compiled separately for each edition then something like the following could work (given that there will be such a cfg value to conditionally compile on to begin with):
```rust
#[cfg_attr(not(edition = "2015"), repr(align(4)))]
pub struct in6_addr {
    pub s6_addr: [u8; 16],
    #[cfg(edition = "2015")]
    __align: [u32; 0],
}
```

Another issue that discusses using newer rustc features in order to improve the libc implementation is #1020. Both implementing that issue and this one would raise the rustc version requirement. Although, the other issue would require rustc 1.19, while this is a lot more extreme as it requires 1.25.

**Comment by alexcrichton** [comment]
Yes unfortunately the version policy for libc isn't strictly documented, but I agree that 1.25 is a bit too recent. Once we decide to upgrade to 1.25 it should be fine to remove all the private `__align` fields and use the attribute instead.

**Comment by faern** [comment]
Idea: Since it does not break the API, and using the `mem::uninitialized()` way of creating the structs should work equally well for the new way of doing alignment. I think one way of making it work today is similar to my brainstorming about editions, but do it with a feature that is off by default:
```rust
#[cfg_attr(feature = "align", repr(align(4)))]
pub struct in6_addr {
    pub s6_addr: [u8; 16],
    #[cfg(not(feature = "align"))]
    __align: [u32; 0],
}
```
Since no crate depending on libc today does activate this, everything will continue to be exactly as now. But as soon as one crate in the dependency chain of some build does add the feature it becomes active for everyone. However, the new layout should not break any existing crates, it should just increase the compiler version required.

So any crate that activates the feature would be equivalent to that same crate doing anything else that makes it require Rust 1.25 or higher. Such as nested import groups or `#[repr(align(x))]` in that crate itself.

Downsides with this approach is that I have not proven the change to be 100% non breaking, even though I think it is. And it also "costs" two `#[cfg(..)]`s per struct that needs alignment.

**Comment by alexcrichton** [comment]
Tha seems like a good solution to me!

**Comment by faern** [comment]
Nice. I'll play around with that and get back with my findings.

**Comment by faern** [comment]
Support for getting rid of `__align` and making libc use `#[repr(align(x))]` was merged in #1044. Just activate the "align" feature of the crate in order to use it.
